### PR TITLE
feat(#220) caching data objects

### PR DESCRIPTION
**Pull request by kadmit**
Implemented a simple library to interact with RabbitMQ

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,173 @@
...
+$publisher->publish(['key' => 'value'], 'name-of-exchange');
+$publisher->publish($objectThatImplementsJsonSerializable, 'name-of-exchange', 'name-of-routing-key');
+$publisher->publish($objectThatImplementsJsonable, 'name-of-exchange', 'name-of-routing-key', [
+    'custom-header-name' => 'header-value'
+]);
+```
+
+### Consuming messages
+
+The lib contains command for consuming messages
```
Comment (DanBondarenko) [inline README.md:155]: **[non-blocker]** could add example of how the library’s consumer is intended to be used

File: config/amqp.php
Comment (DanBondarenko) [inline config/amqp.php:1]: **[education]** is this file necessary here? I would have assumed it’s only required when the library is used as a dependency

File: src/Publisher.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     * @throws Exception
+     */
+    public function setupExchanges(): void
+    {
+        foreach ($this->getProperty('publisher.exchanges', []) as $exchangeConfig) {
+            $this->declareExchange($exchangeConfig['exchange']);
+        }
+    }
+
+    private function declareExchange(string $exchange): void
```
Comment (DanBondarenko) [inline src/Publisher.php]: **[non-blocker]**

```suggestion
    private function declareExchangeIfNotDeclared(string $exchange): void
```

File: src/Consumer.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        );
+
+        foreach ($queueConfig['bindings'] as $bindingConfig) {
+            $this->getChannel()->queue_bind(
+                $queueConfig['queue'],
+                $bindingConfig['exchange'],
+                $bindingConfig['routingKey'] ?? '',
+            );
+        }
+    }
```
Comment (DanBondarenko) [inline src/Consumer.php:145]: **[blocker]** Isn’t the same queue being redeclared everytime a message is consumed? I imagine that might cause issues, but I might be mistaken.

I would validate this by creating a test case that calls `Consumer->consume()` several times and verifies that messages in the queue are still retrievable and no messages are getting lost.

**Review by DanBondarenko — COMMENTED** [review]
@kadmit Hi, thank you for the PR! Overall, looks very nice, but I noticed a potential issue in `Consumer.php`. It might not end up being serious, but just wanted to confirm.

Also, the PR references [#101](https://github.com/keeleinstituut/tv-tolkevarav/issues/101), but I don’t understand how this code is connected to that issue. Are you sure it’s the right issue?

File: config/amqp.php
Comment (kadmit) [reply config/amqp.php:1]: @DanBondarenko The file is needed to configure the consumer/publisher in each application that will use the lib, it's more convenient when you have the template of the config file that's why I decided to add it. I also added `amqp-sample.php` to show all possible config parameters.

File: src/Consumer.php (hunk 3 above)
Comment (kadmit) [reply src/Consumer.php:145]: @DanBondarenko the queue declaration is called once when the process starts working. I tested the behavior and the declaration of the already existing queue will not end up with cleanup. 

In addition, if you declare the queue with the same name and parameters as the existing queue, RabbitMQ will simply return a confirmation message indicating that the queue already exists. This confirmation message can be useful for verifying that the queue has been properly configured and is ready to use.

**Comment by kadmit** [comment]
@DanBondarenko regarding the reference:
We don't have a separate task for the creation of the lib for data sync, and the main idea was to implement the lib and then add sync of institutions and users that were created by the script. I agree that better to not refer to the task here to prevent confusion. Thank you for pointing this out 🙂

File: src/Consumer.php (hunk 3 above)
Comment (DanBondarenko) [reply src/Consumer.php:145]: A-ha, I see. Sorry for the confusion and thank you for the clarification! :) 

Resolving

File: config/amqp.php
Comment (DanBondarenko) [reply config/amqp.php:1]: Thank you for the explanation!

Resolving

**Review by DanBondarenko — APPROVED** [review]
LGTM

**Comment by DanBondarenko** [comment]
@kadmit I see that you have made some updates to this PR. Would you like me to review these changes or are there still further changes coming?

**Comment by kadmit** [comment]
@DanBondarenko the task is not done yet, so no need to review.
